bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseDynamicStateDescriptionProvider.java

```diff
@@ -0,0 +1,152 @@
+import org.eclipse.smarthome.core.thing.i18n.ChannelTypeI18nLocalizationService;
+import org.eclipse.smarthome.core.thing.type.ChannelType;
+import org.eclipse.smarthome.core.thing.type.ChannelTypeUID;
+import org.eclipse.smarthome.core.thing.type.DynamicStateDescriptionProvider;
+import org.eclipse.smarthome.core.types.StateDescription;
+import org.eclipse.smarthome.core.types.StateDescriptionFragmentBuilder;
+import org.eclipse.smarthome.core.types.StateOption;
+import org.osgi.framework.BundleContext;
+import org.osgi.service.component.ComponentContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Deactivate;
+
+/**
+ * The {@link BaseDynamicStateDescriptionProvider} provides a base implementation for the
+ * {@link DynamicStateDescriptionProvider}.
+ * <p>
+ * It provides localized state patterns and dynamic state options while leaving other state description fields as
+ * original. Therefore the inheriting class has to request the reference for the
+ * {@link ChannelTypeI18nLocalizationService} on its own.
+ *
+ * @author Christoph Weitkamp - Initial contribution
+ */
+@NonNullByDefault
+public abstract class BaseDynamicStateDescriptionProvider implements DynamicStateDescriptionProvider {
+
+    private @NonNullByDefault({}) BundleContext bundleContext;
+    protected @NonNullByDefault({}) ChannelTypeI18nLocalizationService channelTypeI18nLocalizationService;
+
+    private final Map<ChannelUID, @Nullable String> channelPatternMap = new ConcurrentHashMap<>();
+    private final Map<ChannelUID, @Nullable List<StateOption>> channelOptionsMap = new ConcurrentHashMap<>();
+
+    /**
+     * For a given channel UID, set a pattern that should be used for the channel, instead of the one defined statically
+     * in the {@link ChannelType}.
+     *
+     * @param channelUID the channel UID of the channel
+     * @param pattern a pattern
+     */
+    public void setStatePattern(ChannelUID channelUID, String pattern) {
+        channelPatternMap.put(channelUID, pattern);
+    }
+
+    /**
+     * For a given channel UID, set a {@link List} of {@link StateOption}s that should be used for the channel, instead
+     * of the one defined statically in the {@link ChannelType}.
+     *
+     * @param channelUID the channel UID of the channel
+     * @param options a {@link List} of {@link StateOption}s
+     */
+    public void setStateOptions(ChannelUID channelUID, List<StateOption> options) {
+        channelOptionsMap.put(channelUID, options);
+    }
+
+    @Override
+    public @Nullable StateDescription getStateDescription(Channel channel, @Nullable StateDescription original,
+            @Nullable Locale locale) {
+        // can be overridden by subclasses
+        ChannelUID channelUID = channel.getUID();
+        String pattern = channelPatternMap.get(channelUID);
+        List<StateOption> options = channelOptionsMap.get(channelUID);
+        if (pattern == null && options == null) {
```

Comment:
This suggests both `setStatePattern` and `setStateOptions` need to be set. Given the javadoc of `setStatePattern` that if not set the default pattern is used I would suspect if `pattern` is null here first `original.getPattern()` should be used and checked if that isn't null either?

Reply:
I am not sure if I get your point. The `StateDescriptionProvider` only should return an new `StateDescription` instance if at least one values is set. In general the `StateDescriptionService` is responsible for merging it with the original `StateDescription`, so we do not need not take care of that here.

Reply:
Ok. I don't know what I was thinking here, maybe I shouldn't write code when tired :smile:. This code seems fine. So I'll have to search further why my other changes didn't work...

Reply:
Never mind. I know that :laugh:.

Do you already created a PR (or have a piece of code) to point me to? I can have a look too. And I have some idea related to the component declaration of a `StateDescriptionProvider` which I like to discuss with you. It does not really fit in here.

Reply:
I haven't created a pr, but the code is here: https://github.com/Hilbrand/openhab2-addons/commits/spotif_base_dynamic
I tested it locally, but not a lot. It doesn't show the dynamic lists anymore, so I started debugging, but not a lot. Interested in your idea.

Reply:
Thanks for the link. See my https://github.com/Hilbrand/openhab2-addons/commit/9c4f586cf6bb64e3cd0c229290a5641e38113521 over there.